From 85914915f543ee3ccf1ea0f826f3af0c1d59789f Mon Sep 17 00:00:00 2001
From: Andre Vehreschild <vehre@gmx.de>
Date: Mon, 28 Apr 2025 09:08:06 +0200
Subject: [PATCH] Use COMM_TEAM instead of COMM_WORLD.

Revert changes to CAF_COMM_WORLD and introduce CAF_COMM_TEAM.
Let all communication happen on CAF_COMM_WORLD and map the
image indexes accordingly. Use CAF_COMM_TEAM for all collective
ops.
---
 src/runtime-libraries/mpi/mpi_caf.c | 155 ++++++++++++++--------------
 1 file changed, 78 insertions(+), 77 deletions(-)

diff --git a/src/runtime-libraries/mpi/mpi_caf.c b/src/runtime-libraries/mpi/mpi_caf.c
index 500a4316..c1599af3 100644
--- a/src/runtime-libraries/mpi/mpi_caf.c
+++ b/src/runtime-libraries/mpi/mpi_caf.c
@@ -358,11 +358,12 @@ typedef struct gfc_max_dim_descriptor_t
 
 char err_buffer[MPI_MAX_ERROR_STRING];
 
-#ifdef GCC_GE_16
 /* All CAF runtime calls should use this comm instead of MPI_COMM_WORLD for
  * interoperability purposes. */
-MPI_Comm *CAF_COMM_WORLD_store;
-#define CAF_COMM_WORLD (*CAF_COMM_WORLD_store)
+MPI_Comm CAF_COMM_WORLD;
+
+#ifdef GCC_GE_16
+#define CAF_COMM_TEAM current_team->team_list_elem->communicator
 
 typedef struct caf_teams_list
 {
@@ -390,10 +391,10 @@ typedef struct caf_team_stack_node
 
 static caf_teams_list_t *teams_list = NULL;
 static caf_team_stack_node_t *current_team = NULL, *initial_team;
+
 #else
-/* All CAF runtime calls should use this comm instead of MPI_COMM_WORLD for
- * interoperability purposes. */
-MPI_Comm CAF_COMM_WORLD;
+
+#define CAF_COMM_TEAM CAF_COMM_WORLD
 
 typedef struct caf_teams_list
 {
@@ -1527,9 +1528,6 @@ PREFIX(init)(int *argc, char ***argv)
     /* Flag rc as unused, because conditional compilation.  */
     int ierr = 0, i = 0, j = 0, rc __attribute__((unused)), prov_lev = 0;
     int is_init = 0, prior_thread_level = MPI_THREAD_MULTIPLE;
-#ifdef GCC_GE_16
-    MPI_Comm tmp_world;
-#endif
 
     ierr = MPI_Initialized(&is_init);
     chk_err(ierr);
@@ -1576,12 +1574,8 @@ PREFIX(init)(int *argc, char ***argv)
 
     /* Duplicate MPI_COMM_WORLD so that no CAF internal functions use it.
      * This is critical for MPI-interoperability. */
-#ifdef GCC_GE_16
-    rc = MPI_Comm_dup(MPI_COMM_WORLD, &tmp_world);
-    CAF_COMM_WORLD_store = &tmp_world;
-#else
     rc = MPI_Comm_dup(MPI_COMM_WORLD, &CAF_COMM_WORLD);
-#endif
+
 #ifdef WITH_FAILED_IMAGES
     flag = (MPI_SUCCESS == rc);
     rc = MPIX_Comm_agree(MPI_COMM_WORLD, &flag);
@@ -1625,8 +1619,8 @@ PREFIX(init)(int *argc, char ***argv)
     /* BEGIN teams preparations. */
 #ifdef GCC_GE_16
     teams_list = (caf_teams_list_t *)malloc(sizeof(caf_teams_list_t));
-    *teams_list = (struct caf_teams_list){tmp_world, -1, caf_this_image, NULL};
-    CAF_COMM_WORLD_store = &teams_list->communicator;
+    *teams_list
+        = (struct caf_teams_list){CAF_COMM_WORLD, -1, caf_this_image, NULL};
     current_team
         = (caf_team_stack_node_t *)malloc(sizeof(caf_team_stack_node_t));
     *current_team = (struct caf_team_stack_node){teams_list, NULL};
@@ -2162,8 +2156,8 @@ void PREFIX(register)(size_t size, caf_register_t type, caf_token_t *token,
         mpi_token = (mpi_caf_token_t *)(*token);
         p = TOKEN(mpi_token);
         mem = desc->base_addr;
-        ierr = MPI_Win_create(mem, actual_size, 1, MPI_INFO_NULL,
-                              CAF_COMM_WORLD, p);
+        ierr = MPI_Win_create(mem, actual_size, 1, MPI_INFO_NULL, CAF_COMM_TEAM,
+                              p);
         chk_err(ierr);
         CAF_Win_lock_all(*p);
         mpi_token->owning_memory = 0;
@@ -2194,15 +2188,15 @@ void PREFIX(register)(size_t size, caf_register_t type, caf_token_t *token,
         p = TOKEN(mpi_token);
 
 #if MPI_VERSION >= 3
-        ierr = MPI_Win_allocate(actual_size, 1, MPI_INFO_NULL, CAF_COMM_WORLD,
+        ierr = MPI_Win_allocate(actual_size, 1, MPI_INFO_NULL, CAF_COMM_TEAM,
                                 &mem, p);
         chk_err(ierr);
         CAF_Win_lock_all(*p);
 #else
         ierr = MPI_Alloc_mem(actual_size, MPI_INFO_NULL, &mem);
         chk_err(ierr);
-        ierr = MPI_Win_create(mem, actual_size, 1, MPI_INFO_NULL,
-                              CAF_COMM_WORLD, p);
+        ierr = MPI_Win_create(mem, actual_size, 1, MPI_INFO_NULL, CAF_COMM_TEAM,
+                              p);
         chk_err(ierr);
 #endif // MPI_VERSION
 #ifdef GCC_GE_16
@@ -2415,7 +2409,7 @@ PREFIX(deregister)(caf_token_t *token, int *stat, char *errmsg,
     /* Sync all images only, when deregistering the token. Just freeing the
      * memory needs no sync. */
 #ifdef WITH_FAILED_IMAGES
-    ierr = MPI_Barrier(CAF_COMM_WORLD);
+    ierr = MPI_Barrier(CAF_COMM_TEAM);
     chk_err(ierr);
 #else
     PREFIX(sync_all)(NULL, NULL, 0);
@@ -5742,13 +5736,13 @@ get_from_self(caf_token_t token, const gfc_descriptor_t *opt_src_desc,
 }
 
 bool
-team_translate(int *remote_image, int *this_image, caf_token_t token,
-               int image_index, caf_team_t *team, int *team_number, int *stat)
+team_translate(int *remote_image, int *this_image,
+               caf_token_t token __attribute__((unused)), int image_index,
+               caf_team_t *team, int *team_number, int *stat)
 {
-  MPI_Group current_team_group, win_group;
-  int ierr;
-  int trans_ranks[2];
-  MPI_Comm comm = CAF_COMM_WORLD;
+  MPI_Group world_group, team_group, remote_group;
+  int ierr, trans, team_id = current_team->team_list_elem->team_id;
+  MPI_Comm remote_comm = MPI_COMM_NULL;
 
 #ifdef GCC_GE_16
   if (team)
@@ -5762,7 +5756,8 @@ team_translate(int *remote_image, int *this_image, caf_token_t token,
                          (*(caf_teams_list_t **)team)->team_id);
       return false;
     }
-    comm = (*(caf_teams_list_t **)team)->communicator;
+    remote_comm = (*(caf_teams_list_t **)team)->communicator;
+    team_id = (*(caf_teams_list_t **)team)->team_id;
   }
   else if (team_number)
   {
@@ -5770,7 +5765,8 @@ team_translate(int *remote_image, int *this_image, caf_token_t token,
     for (caf_team_stack_node_t *cur = current_team; cur; cur = cur->parent)
       if (cur->team_list_elem->team_id == *team_number)
       {
-        comm = cur->team_list_elem->communicator;
+        remote_comm = cur->team_list_elem->communicator;
+        team_id = cur->team_list_elem->team_id;
         found = true;
         break;
       }
@@ -5783,20 +5779,30 @@ team_translate(int *remote_image, int *this_image, caf_token_t token,
   }
 #endif
 
-  ierr = MPI_Comm_group(comm, &current_team_group);
+  ierr = MPI_Comm_group(CAF_COMM_WORLD, &world_group);
   chk_err(ierr);
-  ierr = MPI_Win_get_group(*TOKEN(token), &win_group);
+  if (remote_comm == MPI_COMM_NULL)
+    ierr = MPI_Comm_group(CAF_COMM_TEAM, &remote_group);
+  else
+    ierr = MPI_Comm_group(remote_comm, &remote_group);
   chk_err(ierr);
-  ierr = MPI_Group_translate_ranks(current_team_group, 2,
-                                   (int[]){image_index - 1, mpi_this_image},
-                                   win_group, trans_ranks);
+  ierr = MPI_Comm_group(CAF_COMM_TEAM, &team_group);
   chk_err(ierr);
-  *remote_image = trans_ranks[0];
-  *this_image = trans_ranks[1];
-  ierr = MPI_Group_free(&current_team_group);
+  trans = image_index - 1;
+  ierr = MPI_Group_translate_ranks(team_group, 1, &mpi_this_image, world_group,
+                                   this_image);
   chk_err(ierr);
-  ierr = MPI_Group_free(&win_group);
+  ierr = MPI_Group_translate_ranks(remote_group, 1, &trans, world_group,
+                                   remote_image);
+  chk_err(ierr);
+  ierr = MPI_Group_free(&remote_group);
+  chk_err(ierr);
+  ierr = MPI_Group_free(&team_group);
   chk_err(ierr);
+  ierr = MPI_Group_free(&world_group);
+  chk_err(ierr);
+  dprint("this: %d -> %d, rmt: %d -> %d on team %d.\n", caf_this_image,
+         *this_image + 1, image_index, *remote_image + 1, team_id);
   return true;
 }
 
@@ -5889,7 +5895,7 @@ PREFIX(get_from_remote)(caf_token_t token, const gfc_descriptor_t *opt_src_desc,
   msg->transfer_size = dst_size;
   msg->opt_charlen = opt_src_charlen ? *opt_src_charlen : 0;
   msg->win = *TOKEN(token);
-  msg->dest_image = mpi_this_image;
+  msg->dest_image = this_image;
   msg->dest_tag = CAF_CT_TAG + 1;
   msg->dest_opt_charlen = opt_dst_charlen ? *opt_dst_charlen : 1;
   msg->flags = (opt_dst_desc ? CT_DST_HAS_DESC : 0)
@@ -5932,13 +5938,12 @@ PREFIX(get_from_remote)(caf_token_t token, const gfc_descriptor_t *opt_src_desc,
         caf_runtime_error("Unable to allocate memory "
                           "for internal buffer in get_from_remote().");
     }
-    dprint("waiting to receive %zd bytes from %d.\n", dst_size,
-           image_index - 1);
-    ierr = MPI_Recv(t_buff, dst_size, MPI_BYTE, image_index - 1, msg->dest_tag,
+    dprint("waiting to receive %zd bytes from %d.\n", dst_size, remote_image);
+    ierr = MPI_Recv(t_buff, dst_size, MPI_BYTE, remote_image, msg->dest_tag,
                     CAF_COMM_WORLD, MPI_STATUS_IGNORE);
     chk_err(ierr);
     dprint("received %zd bytes as requested from %d.\n", dst_size,
-           image_index - 1);
+           remote_image);
     // dump_mem("get_from_remote", t_buff, dst_size);
     memcpy(*dst_data, t_buff, dst_size);
 
@@ -6276,7 +6281,7 @@ PREFIX(send_to_remote)(caf_token_t token, gfc_descriptor_t *opt_dst_desc,
   msg->transfer_size = src_size;
   msg->opt_charlen = opt_src_charlen ? *opt_src_charlen : 0;
   msg->win = *TOKEN(token);
-  msg->dest_image = mpi_this_image;
+  msg->dest_image = this_image;
   msg->dest_tag = CAF_CT_TAG + 1;
   msg->dest_opt_charlen = opt_dst_charlen ? *opt_dst_charlen : 1;
   msg->flags = (opt_dst_desc ? CT_DST_HAS_DESC : 0)
@@ -6317,13 +6322,13 @@ PREFIX(send_to_remote)(caf_token_t token, gfc_descriptor_t *opt_dst_desc,
 
   {
     char c;
-    dprint("waiting to receive %d bytes from %d on tag %d.\n", 1, image_index,
+    dprint("waiting to receive 1 byte from %d on tag %d.\n", remote_image,
            msg->dest_tag);
-    ierr = MPI_Recv(&c, 1, MPI_BYTE, image_index - 1, msg->dest_tag,
+    ierr = MPI_Recv(&c, 1, MPI_BYTE, remote_image, msg->dest_tag,
                     CAF_COMM_WORLD, MPI_STATUS_IGNORE);
     chk_err(ierr);
-    dprint("received %d bytes as requested from %d on tag %d.\n", 1,
-           image_index, msg->dest_tag);
+    dprint("received 1 byte as requested from %d on tag %d.\n", remote_image,
+           msg->dest_tag);
   }
 
   if (running_accesses == rat)
@@ -6383,9 +6388,7 @@ PREFIX(transfer_between_remotes)(
     caf_team_t *dst_team, int *dst_team_number, caf_team_t *src_team,
     int *src_team_number)
 {
-  MPI_Group current_team_group, win_group;
   int ierr, this_image, src_remote_image, dst_remote_image;
-  int trans_ranks[3];
   bool free_msg;
   ct_msg_t *full_msg, *dst_msg;
   struct transfer_msg_data_t *tmd;
@@ -6545,7 +6548,7 @@ PREFIX(transfer_between_remotes)(
   dst_msg->transfer_size = src_size;
   dst_msg->opt_charlen = opt_src_charlen ? *opt_src_charlen : 0;
   dst_msg->win = *TOKEN(dst_token);
-  dst_msg->dest_image = mpi_this_image;
+  dst_msg->dest_image = this_image;
   dst_msg->dest_tag = CAF_CT_TAG + 1;
   dst_msg->dest_opt_charlen = opt_dst_charlen ? *opt_dst_charlen : 1;
   dst_msg->flags
@@ -6588,15 +6591,15 @@ PREFIX(transfer_between_remotes)(
 
   {
     char c;
-    dprint("waiting to receive %d bytes from %d on tag %d.\n", 1,
-           dst_image_index, dst_msg->dest_tag);
-    ierr = MPI_Recv(&c, 1, MPI_BYTE, dst_image_index - 1, dst_msg->dest_tag,
+    dprint("waiting to receive 1 byte from %d on tag %d.\n", dst_remote_image,
+           dst_msg->dest_tag);
+    ierr = MPI_Recv(&c, 1, MPI_BYTE, dst_remote_image, dst_msg->dest_tag,
                     CAF_COMM_WORLD, MPI_STATUS_IGNORE);
     chk_err(ierr);
     if (dst_stat)
       *dst_stat = c;
-    dprint("received %d bytes as requested from %d on tag %d.\n", 1,
-           dst_image_index, dst_msg->dest_tag);
+    dprint("received 1 byte as requested from %d on tag %d.\n",
+           dst_remote_image, dst_msg->dest_tag);
   }
 
   if (running_accesses == rat)
@@ -9509,19 +9512,19 @@ internal_co_reduce(MPI_Op op, gfc_descriptor_t *source, int result_image,
     if (result_image == 0)
     {
       ierr = MPI_Allreduce(MPI_IN_PLACE, source->base_addr, size, datatype, op,
-                           CAF_COMM_WORLD);
+                           CAF_COMM_TEAM);
       chk_err(ierr);
     }
     else if (result_image == caf_this_image)
     {
       ierr = MPI_Reduce(MPI_IN_PLACE, source->base_addr, size, datatype, op,
-                        result_image - 1, CAF_COMM_WORLD);
+                        result_image - 1, CAF_COMM_TEAM);
       chk_err(ierr);
     }
     else
     {
       ierr = MPI_Reduce(source->base_addr, NULL, size, datatype, op,
-                        result_image - 1, CAF_COMM_WORLD);
+                        result_image - 1, CAF_COMM_TEAM);
       chk_err(ierr);
     }
     if (ierr)
@@ -9543,19 +9546,19 @@ internal_co_reduce(MPI_Op op, gfc_descriptor_t *source, int result_image,
                         + array_offset_sr * GFC_DESCRIPTOR_SIZE(source));
     if (result_image == 0)
     {
-      ierr = MPI_Allreduce(MPI_IN_PLACE, sr, 1, datatype, op, CAF_COMM_WORLD);
+      ierr = MPI_Allreduce(MPI_IN_PLACE, sr, 1, datatype, op, CAF_COMM_TEAM);
       chk_err(ierr);
     }
     else if (result_image == caf_this_image)
     {
       ierr = MPI_Reduce(MPI_IN_PLACE, sr, 1, datatype, op, result_image - 1,
-                        CAF_COMM_WORLD);
+                        CAF_COMM_TEAM);
       chk_err(ierr);
     }
     else
     {
       ierr = MPI_Reduce(sr, NULL, 1, datatype, op, result_image - 1,
-                        CAF_COMM_WORLD);
+                        CAF_COMM_TEAM);
       chk_err(ierr);
     }
     if (ierr)
@@ -9618,13 +9621,13 @@ PREFIX(co_broadcast)(gfc_descriptor_t *a, int source_image, int *stat,
     if (datatype == MPI_BYTE)
     {
       ierr = MPI_Bcast(a->base_addr, size * GFC_DESCRIPTOR_SIZE(a), datatype,
-                       source_image - 1, CAF_COMM_WORLD);
+                       source_image - 1, CAF_COMM_TEAM);
       chk_err(ierr);
     }
     else if (datatype != MPI_CHARACTER)
     {
       ierr = MPI_Bcast(a->base_addr, size, datatype, source_image - 1,
-                       CAF_COMM_WORLD);
+                       CAF_COMM_TEAM);
       chk_err(ierr);
     }
     else
@@ -9633,13 +9636,13 @@ PREFIX(co_broadcast)(gfc_descriptor_t *a, int source_image, int *stat,
       if (caf_this_image == source_image)
         a_length = strlen(a->base_addr);
       /* Broadcast the string lenth */
-      ierr = MPI_Bcast(&a_length, 1, MPI_INT, source_image - 1, CAF_COMM_WORLD);
+      ierr = MPI_Bcast(&a_length, 1, MPI_INT, source_image - 1, CAF_COMM_TEAM);
       chk_err(ierr);
       if (ierr)
         goto error;
       /* Broadcast the string itself */
       ierr = MPI_Bcast(a->base_addr, a_length, datatype, source_image - 1,
-                       CAF_COMM_WORLD);
+                       CAF_COMM_TEAM);
       chk_err(ierr);
     }
 
@@ -9668,7 +9671,7 @@ PREFIX(co_broadcast)(gfc_descriptor_t *a, int source_image, int *stat,
     void *sr = (void *)((char *)a->base_addr
                         + array_offset * GFC_DESCRIPTOR_SIZE(a));
 
-    ierr = MPI_Bcast(sr, 1, datatype, source_image - 1, CAF_COMM_WORLD);
+    ierr = MPI_Bcast(sr, 1, datatype, source_image - 1, CAF_COMM_TEAM);
     chk_err(ierr);
 
     if (ierr)
@@ -10426,7 +10429,7 @@ PREFIX(form_team)(int team_id, caf_team_t *team, int *new_index, int *stat,
   static const char *negative_new_index
       = "FORM TEAM: NEW_INDEX= shall be a positive unique integer";
   caf_teams_list_t *new_team;
-  MPI_Comm current_comm = CAF_COMM_WORLD;
+  MPI_Comm current_comm = CAF_COMM_TEAM;
   int ierr;
   int key = new_index ? *new_index : caf_this_image;
 
@@ -10504,13 +10507,12 @@ PREFIX(change_team)(caf_team_t team, int *stat, char *errmsg,
       = (caf_team_stack_node_t){provisioned_team, NULL, current_team};
 
   current_team = new_current_team;
-  CAF_COMM_WORLD_store = &current_team->team_list_elem->communicator;
-  ierr = MPI_Comm_rank(CAF_COMM_WORLD, &mpi_this_image);
+  ierr = MPI_Comm_rank(CAF_COMM_TEAM, &mpi_this_image);
   chk_err(ierr);
   caf_this_image = mpi_this_image + 1;
-  ierr = MPI_Comm_size(CAF_COMM_WORLD, &caf_num_images);
+  ierr = MPI_Comm_size(CAF_COMM_TEAM, &caf_num_images);
   chk_err(ierr);
-  ierr = MPI_Barrier(CAF_COMM_WORLD);
+  ierr = MPI_Barrier(CAF_COMM_TEAM);
   chk_err(ierr);
   dprint("changed to team %d.\n", current_team->team_list_elem->team_id);
 }
@@ -10638,7 +10640,7 @@ PREFIX(end_team)(int *stat, char *errmsg, charlen_t errmsg_len)
   if (stat)
     *stat = 0;
 
-  ierr = MPI_Barrier(CAF_COMM_WORLD);
+  ierr = MPI_Barrier(CAF_COMM_TEAM);
   chk_err(ierr);
   if (current_team->parent == NULL)
   {
@@ -10671,11 +10673,10 @@ PREFIX(end_team)(int *stat, char *errmsg, charlen_t errmsg_len)
   }
   current_team = current_team->parent;
 
-  CAF_COMM_WORLD_store = &current_team->team_list_elem->communicator;
-  ierr = MPI_Comm_rank(CAF_COMM_WORLD, &mpi_this_image);
+  ierr = MPI_Comm_rank(CAF_COMM_TEAM, &mpi_this_image);
   chk_err(ierr);
   caf_this_image = mpi_this_image + 1;
-  ierr = MPI_Comm_size(CAF_COMM_WORLD, &caf_num_images);
+  ierr = MPI_Comm_size(CAF_COMM_TEAM, &caf_num_images);
   chk_err(ierr);
   free(ending_team);
   dprint("switched to team %d.\n", current_team->team_list_elem->team_id);
